### DISCUSSION TITLE
unifying generators (id, version, value generation)

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#5349] [Design Proposals]
Perhaps I'm wrong, or perhaps it's simply not worth the effort, but I feel like `VersionJavaType` is a bit of a hangover from ancient days, and that there should be some way to unify version generation with the whole `ValueGeneration` infrastructure.

In particular, the ugly `@Source` annotation, and the `DbTimestampType` (which is deprecated, but still underlies `@Source(DB)`) stick out like sore thumbs.

@beikov wrote in `DbTimestampType` that we should:
```java
Use {@link org.hibernate.tuple.ValueGeneration} instead
```
but as far as I can tell it's not *actually* possible at present to use `ValueGeneration` to implement `@Source`. 

Nor can I see any way to achieve the effect of `@Source` using `@Generated` or `@GeneratedColumn`.  

Am I missing something there? I would love to deprecate `@Source` or reimplement it as a `@ValueGenerationType`.

Thoughts?

**Comment by beikov** [comment]
Using the annotation `@CurrentTimestamp` which uses `CurrentTimestampGeneration` would be the "equivalent" in the `ValueGeneration` realm. AFAIR this can only be applied through annotations though. So to support the special named type `dbtimestamp`, I assume we'd have to special case this type name somewhere in a binder and use the `timestamp` type instead + apply the `CurrentTimestampGeneration`.

**Reply by gavinking** [reply]
> Using the annotation `@CurrentTimestamp` which uses `CurrentTimestampGeneration` would be the "equivalent" in the `ValueGeneration` realm.

Right, what I'm suggesting it to integrate `ValueGeneration` with the optimistic locking stuff. Currently you can't use a value generated via `ValueGeneration` to do optimistic locking.

**Comment by beikov** [comment]
The `VersionJavaType` stuff is what I introduced as replacement for `VersionType`. This is used for optimistic version generation. The logic was only dependent on the java type and since we wanted to get rid of the special `BasicType` implementations anyway, the move to the `JavaType` hierarchy for this was natural.

**Reply by gavinking** [reply]
Sure, and that's all perfectly reasonable. The problem long predates the migration of `VersionType` to `VersionJavaType`.

**Comment by gavinking** [comment]
So after today's work, this should now be straightforward. We can make the move anytime we like, since all the machinery is there.

**Reply by gavinking** [reply]
In fact, this is eventually going to go even further that was originally planned: I have also corralled `IdentifierGenerator` underneath `InMemoryGenerator`, and soon we should be able to use any `InMemoryGenerator` to generate ids.

So this is turning into unification of id, version, and "value" generation. Eventually we will have just one thing. (Probably not for 6.2, due to concerns about compatibility of SPI interfaces.)

**Reply by gavinking** [reply]
@sebersole how do we feel about changing the return types of methods on `IdentifierGeneratorFactory`? (Widening them.)

**Reply by gavinking** [reply]
> @sebersole how do we feel about changing the return types of methods on `IdentifierGeneratorFactory`? (Widening them.)

What I'm asking is: is it too breaky to change the return types from `IdentifierGenerator` to `InMemoryGenerator`? I don't really see this as a big deal, but perhaps you disagree.

**Reply by sebersole** [reply]
Not all id generators are in-memory, so what's the plan there?  Perhaps you mean `Generator` (which is the super for `InMemoryGenerator`)?

It also affects integrations.  E.g. we were just talking about `IdentifierGeneratorFactory` in regards to HR.  Such a change forces changes to those integrations.

Otherwise, I'm generally fine with the change.  It would actually help addressing the intention behind https://github.com/hibernate/hibernate-orm/discussions/3823

**Reply by gavinking** [reply]
Yeah right, so that's the bit that's still missing thing here. The issue is `PostInsertIdentifierGenerator` and its subtypes, in particular `IdentityGenerator`, which work by returning a magical value which tells `AbstractSaveEventListener` to do something special. And then there's other spots in the code which do what is essentially an `instanceof PostInsertIdentifierGenerator` and then some special thing. This was a very bad implementation I came up 20 years ago that I already knew was bad at the time I did it and was meant to be temporary. (The point is that a `PostInsertIdentifierGenerator` isn't really an `IdentityGenerator` it never was. It was an incorrect model.)

Now, sometime later:

- Somebody (you, Steve, I think) introduced the `InsertGeneratedIdentifierDelegate` abstraction which sort-of *somewhat* improved on my bad solution (it at least made it more flexible). But `PostInsertIdentifierGenerator` were still (not!) `IdentityGenerator`s, so that was still broken. 
- And somebody else (Gunnar, I believe) introduced the `ValueGeneration` stuff which handled regular field values in a way that was in certain aspects better and in some other aspects worse.

What needs to happen now is: the `PostInsertIdentifierGenerator`s, which ultimately descend from `InMemoryGenerator`, should now become `InDatabaseGenerator`s.  That's slightly-easier said than done because they have all this `InsertGeneratedIdentifierDelegate` machinery which is different to the machinery in `InDatabaseGenerator`. 

The good news is, it's probably, with some work, *very complimentary* to what `InDatabaseGenerator` already has, since it has support for stuff like `insert ...returning`. By making these things work together we would make *both* value generation *and* id generation more flexible and elegant. In a sense, `InDatabaseGenerator` has a hardcoded `AbstractSelectingDelegate` in it, and this could be generalized.

The bad news is I barely even had time to *look* at that part of it yet, and I don't know if I will have time for 6.2. I had decided that I was going to leave that for a later release.

But yes, exactly as you say, ultimately it means widening some internal types from `InMemoryGenerator` to just `Generator`. If you think we should do that much *now*, then, yeah, I guess you're probably right.

**Reply by gavinking** [reply]
By the way @sebersole, I just noticed that `SelectGenerator` has a parameter named `"key"`. This means that when we move it over to the `InDatabaseGenerator` branch, we're going to need to generalize the `configure()` method away from `IdentifierGenerator`.
I think the easiest way to handle that is going to be to un-deprecate the `Configurable` interface, so that `SelectGenerator` would implement `InDatabaseGenerator, Configurable`.

**Reply by gavinking** [reply]
> Somebody (you, Steve, I think) introduced the `InsertGeneratedIdentifierDelegate` abstraction

Sorry, I meant `IdentityColumnSupport` of course.

**Reply by sebersole** [reply]
> The issue is PostInsertIdentifierGenerator and its subtypes, in particular IdentityGenerator, which work by returning a magical value which tells AbstractSaveEventListener to do something special

That's just in a very special case where the PersistentContext is extended ("conversation scope") and the persisted entity is using identity generation.  In this case, we actually should not do the insert during "intermediate transations" but we also need *A* value for tracking in the PersistentContext.  Happy enough to hear other suggestions on how to handle that ;)

**Reply by sebersole** [reply]
> And then there's other spots in the code which do what is essentially an instanceof PostInsertIdentifierGenerator and then some special thing. This was a very bad implementation I came up 20 years ago that I already knew was bad at the time I did it and was meant to be temporary. (The point is that a PostInsertIdentifierGenerator isn't really an IdentityGenerator it never was. It was an incorrect model.)

The better option, which I've been using lately, is to check `EntityMutationTarget#getIdentityInsertDelegate` to decide this

**Reply by sebersole** [reply]
Also, check how `InsertCoordinator` handles the case of identity insertions.  That's the only way identity insertions happen anymore, so I think this line of comments is moot.

**Reply by sebersole** [reply]
I'll need to sit down and consider all of this.

Regarding the need to drop `#configure` iiuc, I 1000% agree.  Again, see https://github.com/hibernate/hibernate-orm/discussions/3823.  Dropping `#configure` is one of the main intentions there

**Reply by sebersole** [reply]
> Also, check how `InsertCoordinator` handles the case of identity insertions. That's the only way identity insertions happen anymore, so I think this line of comments is moot.

To clarify here, `InsertCoordinator` handles post-insert generations is a very unique way and does not rely on the IdentifierGenerator at all. It does this via `MutationExecutorPostInsert` / `MutationExecutorPostInsertSingleTable`

**Reply by gavinking** [reply]
> The better option, which I've been using lately, is to check EntityMutationTarget#getIdentityInsertDelegate to decide this

Well what I'm thinking is that `Generator.generatedByDatabase()` is now the blessed Way.

**Reply by gavinking** [reply]
> Dropping `#configure` is one of the main intentions there

Yeah, great, I hate it so much.

**Reply by gavinking** [reply]
> To clarify here, InsertCoordinator handles post-insert generations is a very unique way

Hrm, OK, I will check it out. But whatever, it's probably still going to take at least *some* work to unify that with `GeneratedValuesProcessor`. Though I now think it can be done.

**Reply by gavinking** [reply]
Alright, looks like I have something that is reasonable. There are two rather interesting things about this:

1. I had to repurpose the whole `InsertGeneratedIdentifierDelegate` hierarchy to use the `InDatabaseGenerator` as an intermediary, instead of calling the `IdentityColumnSupport` directly.
2. I had to remove the nexus between generators and `InsertGeneratedIdentifierDelegate`. Now generators have no say in the choice of `InsertGeneratedIdentifierDelegate`, that is up to the persister and the `InsertCoordinator`. At first I was uncomfortable with this but then I realized it's actually much more correct. `SelectGenerator` isn't  a real generator, it doesn't generate the value differently to `IdentityGenerator`, it just hardcodes a different way to *fetch* the generated value. But if generators are now supposed to be used in so many different ways, it doesn't get to make that decision. That's something that really does depend on how it's being called. So this is all good.
3. I had to kill some bad code in `IdentityGeneratorHelper` and use proper types instead. This is much better.
4. I believe I've successfully generated the `PostInsertIdentifierGenerator`s to handle multi-column generated ids. At least in principle.

**Reply by gavinking** [reply]
A remaining task is to fix `StandardTableExporter` so that it can correctly detect and export identity columns, wherever they occur in the mappings. Currently there is a badly-hardcoded special case for a single-column PK. It should be up to the `Column` to know it is an `IDENTITY` column.

**Reply by gavinking** [reply]
Another remaining task is: investigate whether `IdentityColumnSupport.appendIdentitySelectToInsert(String identityColumnName, String insertString)` could be generalized to accept an array of column names. Looks like we could definitely do this on h2, not sure about any other database.

**Reply by gavinking** [reply]
> Looks like we could definitely do this on h2, not sure about any other database.

Well, on postgres it's `returning`, I knew that. And it looks like it's `output` on SQL Server?

**Reply by sebersole** [reply]
> 2\. `SelectGenerator` isn't  a real generator, it doesn't generate the value differently to `IdentityGenerator`

My point earlier was that really post-insert generators are not generators at all

**Reply by gavinking** [reply]
Yes, right, they weren't. Well, they are now :)

**Comment by gavinking** [comment]
So now I've arrived at an interesting question. I've reached the point where an "id generator" is just an `InMemoryGenerator` with `GenerationTiming.INSERT`.

So then what precisely is the difference between an `@IdGeneratorType` annotation and a `@ValueGenerationType` annotation? In principle we really don't need both of them, and in what I pushed yesterday, I let you use a `@ValueGenerationType` to generate ids.

But, since we already have both these annotations, and because I'm disinclined to deprecate the one _we only just added in 6.0_, probably this was a mistake. If you want to use an `InMemoryGenerator` for both things, it's easy enough to define an annotation with is _both_ an `@IdGeneratorType` _and_ a `@ValueGenerationType`. So I think that's the path I'm going to take for now. That is: id generators must be annotated `@IdGeneratorType`, and other field generators must be annotated `@ValueGenerationType`, just like in 6.0.

Thoughts?

**Reply by beikov** [reply]
One thing I'd say that makes id generators different is that they might require a sequence or table i.e. there is some "Exportable" thing related to it.

**Reply by gavinking** [reply]
Yes but in principle so could any other generator ... that we don't currently support that is a (fixable) limitation of the old `ValueGeneration` stuff.

**Reply by gavinking** [reply]
(Indeed, I *want* to see it fixed.)

**Reply by sebersole** [reply]
They might also want an exportable for non-id generation.  This unification ultimately lets us address that

**Reply by sebersole** [reply]
I.e., this might be cool:

```
@Entity 
class ... {
    @Id private Integer id;

    @Basic
    @Generated( strategy = TriggerGenerator.class, ... )
    @TriggerGenerator( ... )
    private String status;
}
```

**Comment by gavinking** [comment]
So I'm now primarily worried about package organization. Related issue is this one: https://hibernate.atlassian.net/browse/HHH-15800

Basically, I want to move `Generator` and friends out of `org.hibernate.tuple`, partly because that package is sort of a mess. I've been through a bunch of options in my head, trying to find a way to do this without breaking. It's very tricky, basically due to the `GenerationTime(ing)` enums, but what I've finally come up with is this:

1. Remove `getGenerationTiming()` from `Generator` (it would remain on the now-legacy `ValueGeneration`) and replace it with two booleans. This makes perfect sense on its own; it was always a bit weird to have an enum for the type `(boolean,boolean)`.
2. Move the new `Generator` interfaces to the new package `org.hibernate.generator`. I think this is important to do.
3. Leave `ValueGeneration`, `GenerationTiming` and friends where they are, in `org.hibernate.tuple` to retain compatibility, though they would now be in a semi-deprecated state.
4. Move concrete implementations of `Generator` to `org.hibernate.generator.internal`. (We could leave inheriting stubs in `org.hibernate.tuple` if you guys think that's important enough to bother with.)
5. Make a final decision as to whether `@Generated` and `@CurrentTimestamp` should be using `annotations.GenerationTime` or `tuple.GenerationTiming`. I had previously decided to standardize them on `tuple.GenerationTiming`, but with this proposal I guess it makes much more sense to reverse this decision, since the whole `tuple` package doesn't have a big future. The thing making this uncomfortable is that we introduced `@CurrentTimestamp` in 6.0, with `timing` given by a `tuple.GenerationTiming`. What I'm thinking now is: can we just call this a "bug" and fix it? I judge the potential breakiness to be quite small. People who have already migrated to 6 probably have not found this annotation yet.

WDYT?

**Reply by gavinking** [reply]
See https://github.com/hibernate/hibernate-orm/pull/5680

**Reply by sebersole** [reply]
> 1. Remove `getGenerationTiming()` from `Generator` (it would remain on the now-legacy `ValueGeneration`) and replace it with two booleans. This makes perfect sense on its own; it was always a bit weird to have an enum for the type `(boolean,boolean)`.

We'll agree to disagree ;)  To me, modeling 3 potential outcomes as 4 options is weird

**Reply by sebersole** [reply]
Overall I agree about the tuple package.  

I've been wanting to tackle that for 6.2, but to me that also includes cleaning `org.hibernate.persister.walking`.  I got most of `org.hibernate.persister.walking` done, but the one remaining contract gets deeply tangled with the tuple package.  

It also, again very unfortunately, gets tangled up with `PropertyMapping` which is another thing I want to get rid of (pretty sure we no longer use that in any meaningful way).

At a high-level, that was all a preliminary attempt at what has become the mapping model but twisted to fit into pre-6.0 codebase.

**Reply by gavinking** [reply]
> We'll agree to disagree ;) To me, modeling 3 potential outcomes as 4 options is weird

Well I added support `UPDATE` to the enums in HHH-15672. https://hibernate.atlassian.net/browse/HHH-15672

I mean, I agree that generation that happens only on `UPDATE` is less *common* than the other options, but there's nothing conceptually wrong wit the idea, it's very easy to support, and surely someone, someday will find it useful.

**Reply by gavinking** [reply]
> It also, again very unfortunately, gets tangled up with `PropertyMapping` which is another thing I want to get rid of (pretty sure we no longer use that in any meaningful way).

Yeah, looks 'bout obsolete to me.

**Reply by sebersole** [reply]
> > We'll agree to disagree ;) To me, modeling 3 potential outcomes as 4 options is weird
> 
> Well I added support `UPDATE` to the enums in HHH-15672. https://hibernate.atlassian.net/browse/HHH-15672
> 
> I mean, I agree that generation that happens only on `UPDATE` is less _common_ than the other options, but there's nothing conceptually wrong wit the idea, it's very easy to support, and surely someone, someday will find it useful.

Yes, you added that VERY recently.  It is so uncommon that no one has asked for that ever as far as I know.

But fine.  You are correct that it is easy to support, usefulness and need aside.
